src/Network/Mattermost/Util.hs

```diff
@@ -132,48 +130,9 @@ buildPath cd endpoint = do
   return $ URI.render uri
 
 
-populateAuth :: ConnectionData -> Maybe Token -> HTTP.Request a -> HTTP.Request a
-populateAuth cd token req =
-  let authHeader = mkAuthHeaders cd token
-  in req { HTTP.rqHeaders = HTTP.rqHeaders req <> authHeader
-         }
-
-
-csrfHeader :: BasicAuth -> [HTTP.Header]
-csrfHeader ba =
-  case baCsrfToken ba of
-    Just t ->
-      [ HTTP.mkHeader (HTTP.HdrCustom "X-CSRF-Token") $ T.unpack t
-      ]
-    Nothing -> []
-
-mkAuthHeaders :: ConnectionData -> Maybe Token -> [HTTP.Header]
-mkAuthHeaders (cd@ConnectionData {cdBasicAuth=Just ba}) (Just t) =
-   mkAuthHeaders cd Nothing <> csrfHeader ba <> [
-     HTTP.cookiesToHeader $ [
-                            HTTP.MkCookie
-                            { HTTP.ckDomain = (T.unpack $ cdHostname cd),
-                              HTTP.ckName = "MMAUTHTOKEN",
-                              HTTP.ckValue = (getTokenString t),
-                              HTTP.ckPath = Nothing,
-                              HTTP.ckComment = Nothing,
-                              HTTP.ckVersion = Nothing
-                            }
-                            ]
-     ]
-
-
-
-mkAuthHeaders _ (Just token) =
-  [HTTP.mkHeader HTTP.HdrAuthorization $ "Bearer " ++ getTokenString token]
-
-
-mkAuthHeaders (ConnectionData {cdBasicAuth=(Just ba)}) Nothing =
-  [HTTP.mkHeader HTTP.HdrAuthorization $ "Basic " ++ (T.unpack $ encodeBasicAuth ba)]
-
-mkAuthHeaders _ _ =
-  []
-
-
-encodeBasicAuth :: BasicAuth -> T.Text
-encodeBasicAuth BasicAuth {baUsername = u, baPassword = p} = encodeBase64 $ T.intercalate (T.pack ":") [u, p]
+populateAuth :: ConnectionData -> Maybe Token -> HTTP.Request_String -> HTTP.Request_String
+populateAuth (ConnectionData {cdConnReqTrans = trans}) token req =
+  trans $ case token of
```

Comment:
Why does `populateAuth` call `trans`? It should not need to.

Reply:
Why not? I added it as central place to add authorization headers. Now it just delegates to the transformation function.
Where would you put the transformation call?

Reply:
Btw, I just realized there is as mistake in this function. Trans has to be called always. Not only if we already know a token. Before authorizing to get a token we already need to authorize against possible other auth schemes

Reply:
> Btw, I just realized there is as mistake in this function. 

Oh wait, my bad. That was nonsense

Reply:
I asked about this because it shouldn't be necessary: `populateAuth` *is* the value of `trans`, and function composition via `.` should suffice at the call site for `setConnectionRequestTransformation` to make sure that all transformations are used, e.g.,
```
setConnectionRequestTransformation cd (populateAuth . otherThing . otherThing2)
```

Reply:
Ok, that would be possible, but then every client would have to configure the request transformer, which would be a breaking change. I wanted to avoid that.

Reply:
Oh, I see. I had assumed all of the request transformations would be factored out (including any existing token-based auth transformations) but I see now that that would potentially cause a problems for clients relying on that behavior. In that case I think the token-based auth that we currently have should be built-in, and any *additional* auth transformations (such as for basic auth) should be provided via the transformation interface you've added. Does that make sense?

Reply:
Since the `ConnectionData` object is usually obtained via `mkConnectionData` or `initConnectionData`, could the current token-based auth be the default provided by those functions?  This would still allow it to be factored out (aside from the default call) and be backward compatible but also allow update by the `setConnectionRequestTransformation`.

Reply:
I was thinking that using the new functionality in this patch for the token authentication runs the risk that it would be easy to forget to include it in a call to `setConnectionRequestTransformation` to keep it enabled. So that's why I was asking above whether we could just keep that one feature enabled as-is and allow further customization via the new API.

Reply:
I tried to define the RequestTransformer as `ConnectionData -> Request -> Request` and then use populateAuth as the default. But i had experienced the problem that the token in connection data was not always populated and didn't look further into it. I think that would be the best design.